model/value_histogram.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2013 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+)
+
+type IntString uint64
+
+func (v IntString) String() string {
+	return fmt.Sprintf("%d", v)
+}
+
+func (v IntString) MarshalJSON() ([]byte, error) {
+	return json.Marshal(v.String())
+}
+
+func (v *IntString) UnmarshalJSON(b []byte) error {
+	if len(b) < 2 || b[0] != '"' || b[len(b)-1] != '"' {
+		return fmt.Errorf("int value must be a quoted string")
+	}
+	i, err := strconv.ParseInt(string(b[1:len(b)-1]), 10, 64)
+	if err != nil {
+		return err
+	}
+	*v = IntString(i)
+	return nil
+}
+
+func (v IntString) Equal(o IntString) bool {
+	return v == o
+}
+
+type FloatString float64
+
+func (v FloatString) String() string {
+	return strconv.FormatFloat(float64(v), 'f', -1, 64)
+}
+
+func (v FloatString) MarshalJSON() ([]byte, error) {
+	return json.Marshal(v.String())
+}
+
+func (v *FloatString) UnmarshalJSON(b []byte) error {
+	if len(b) < 2 || b[0] != '"' || b[len(b)-1] != '"' {
+		return fmt.Errorf("float value must be a quoted string")
+	}
+	f, err := strconv.ParseFloat(string(b[1:len(b)-1]), 64)
+	if err != nil {
+		return err
+	}
+	*v = FloatString(f)
+	return nil
+}
+
+func (v FloatString) Equal(o FloatString) bool {
+	return v == o
+}
+
+type HistogramBucket struct {
+	Boundaries int
+	Lower      FloatString
+	Upper      FloatString
+	Count      IntString
```

Comment:
The count can be a float, actually. Histograms can be so-called "float histograms". (Result of the `rate` function, but it could even happen in instrumentation, as K8s has proven.)

Reply:
Noted, will change all `IntString`s to `FloatString`s, and the corresponding ints in equivalent Mimir structs to floats. The counts will still always be positive right? Or perhaps possibly negative if there's a rate? I guess for simplicity we'll just use float histograms for everything even though they are rarer, since we can cast from int to float but not reverse?

Reply:
I think the `Count` will always be positive (or zero, or NaN). But I don't think this gives us anything. Since we treat it as a float, we would be able to handle negative values, should they show up for some weird reason. (Maybe we allow a delta histogram of two gauge histograms one day, and then the values could be negative…)

And yes, as discussed elsewhere, PromQL is only acting on floats, so it is always acting on float histograms, so all the output from the HTTP query API is always float histograms, even if the float numbers in them all happen to be integral.